### ISSUE TITLE
Integrate memory block merging

### ISSUE THREAD

**Issue by nqpz** [diku-dk/futhark#461]
## Recap

+ In-place lowering: The existing optimisation that we would like to be more general than.
+ Memory block coalescing: The optimisation that looks at `copy`, `concat`, and in-place update statements and tries to merge the left-hand side memory block with the right-hand side memory block.
+ Memory block reuse: The optimisation that tries to let statements reuse memory blocks whose last uses have occured.

## Short status

I have updated the code to work with the current compiler infrastructure and AST; see the https://github.com/diku-dk/futhark/tree/niels-memory-block-merging-final branch.  It passes all our 90+ memory block-specific tests, but it does not handle all our general test programs and benchmarks yet (which means that there are some more things we need to test).  The memory footprint reduction is always non-negative and ranges from 0% to 70%.

I think I can fix the problems mentioned below, but my original timeframe for doing so slipped, so I'm not sure.

## Current compilation and transformation problems

+ [x] Type errors after memory block reuse.  This happens in `tests/futlib_tests/random.fut` and a couple of benchmarks and is related to how the memory block merging modifies a program.  Something goes wrong, some problems seemingly related to IfAttr and constant-size memory blocks, and some seemingly related to the "max trick" employed by memory block reuse.  All of the program transformation happens in the MemoryBlockMerging.MemoryUpdater module, so this is the only place that needs fixing.
+ [x] In the GPU pipeline, the modification of InterestCalib manages to create a program with a memory block size that is not lambda-invariant.  We should probably just restrict memory block merging inside lambdas.
+ [x] In the CPU pipeline, GHC runs out of memory on compilation of InterestCalib.  This is likely due to some silly bug.
+ [x] heston32 compiles but gives a segmentation fault when it is run.  heston64 does not, so this is probably related to how the code calculates the byte size of a memory block.
+ [x] nn-opt-1 and bfs_filt_padded_fused give the wrong results when memory reuse is enabled.  Maybe the interference analysis is lacking in some edge cases.

## Runtime regressions

+ [x] LocVolCalib and nn-opt-4 (at least) run a bit slower when memory reuse is *enabled*.  Even the most basic setting of "only reuse memory if the blocks do not interfere and the block sizes are exactly the same" results in a slower program.  I'm at a bit of a loss here.
+ [ ] lud runs a bit slower when in-place lowering is *disabled*.  The program contains a single in-place statement inside a loop with many iterations.  Memory block coalescing it is a bit conservative wrt. one of its safety conditions (3), and so does not handle it.  But I think it can be made to do so by extending the existing index access analysis.
+ [ ] Not really a regression, but: The coalescing pass merges the two memory blocks in programs like
```
let a = map ... -- mem_a
let b = concat a ... -- mem_b
```
so that both `a` and `b` use `mem_b`, but in the next-to-final ImpGen code, there will still be a copy statement from `mem_b` to `mem_b`.  This is silly, but I have not found a way to do it nicely.  (There are of course also a few edge cases.)

## Code structure

Minor practical matters:

+ [ ] The Futhark.Passes module still contains code for disabling and enabling memory block merging passes.  This should go away if merged into master.
+ [x] There are a lot of debug prints in the code, but they all look the same and should be easy to get rid of.
+ [ ] I extended futhark-bench to also save stderr when using the `--json` flag.  This was an easy way to get the result of `--debugging`, but maybe there is a nicer way.

That's it!  In conclusion, it appears to almost work, but there are a few bugs.  I will try to fix them this week.

**Comment by Athas** [comment]
LGTM

**Comment by nqpz** [comment]
*Status:*

The current code is at branch https://github.com/diku-dk/futhark/commits/niels-memory-block-merging-final.

All benchmarks and tests (including several new tests; see `tests/memory-block-merging`) compile and validate.

On the GPU, all benchmarks except `rodinia/lud.fut` run faster or at the same speed with in-place lowering disabled and memory block merging enabled.

All benchmarks except `parboil/tpacf/tpacf.fut` and `rodinia/srad/srad.fut` have a lower or the same peak memory usage.

**lud**: The in-place lowering optimisation lowers in-place updates in the `lud_perimeter_upper` and `lud_perimeter_lower` functions, while the coalescing algorithm does not, resulting in a small slowdown (84 ms compared to 92 ms in the largest dataset).  The coalescing algorithm notices the potential coalescing, but does not believe it to be safe, since the destination memory in the in-place update is also read from inside the kernel equivalent of the `perimeter` functions. In this case it is safe, though I'm not even sure the in-place lowering optimisation checks that.  I think a proper check could be added to the coalescing algorithm, but I'm not sure how much work it is.

**tpacf and srad**: Until recently, memory block merging never resulted in an increase in peak memory usage, but now it does in these two cases.  They appear to hit upon edge cases like when

1. Allocate a memory block A of size 1.
2. Allocate a memory block B of size 1.
3. Free A.
4. Free B.
5. Allocate a memory block C of size 3.
6. Free C.

with peak memory usage 3 is transformed into

1. Allocate a memory block A of size 3 (now also used instead of C).
2. Allocate a memory block B of size 1.
3. Free B.
4. Free A.

with peak memory usage 4.  Can this maybe have something to do with the new memory allocator?  In any case, I haven't really thought about how to handle it.

Except for these cases it works well, though there are several FIXMEs in the code of varying size.

**Comment by Athas** [comment]
Great work!  Do you have a summary somewhere of the reduction in peak
memory usage for each benchmark?  I am curious.

I think the LUD regression is fine, but I'll see what Cosmin has to day.

I don't think the new allocator is the reason you are seeing peak usage
increases for tpacf and srad, since we don't count the memory that is
stuck in the free list.

Do you think we should merge this?

-- 
\  Troels
/\ Henriksen

**Comment by nqpz** [comment]
There are some numbers for the GPU results here: http://napoleon.hiperfit.dk/~niels/memory-block-merging/gpu01-2018-02-08-futhark-opencl/summary.html -- the apparent slowdowns in some benchmarks are related to those benchmarks having only small datasets, so we can't really say much about those.

My main annoyance with the code right now is that there is a bug (I think) in the simplifier that causes the tpacf benchmark to not be compiled if not for the hack described in the FIXME in `Futhark.Passes`.  It would be nice to get rid of that hack.  I'm not sure which rule in the simplifier is responsible, but I get a type error if I run the entire ExplicitMemory simplifier after performing memory reuse on tpacf.  Or maybe it's the fault of the optimisation?  I'm not sure.

Other than that, the optimisations are still disabled by default, and in-place lowering is still enabled. I think it can be merged.

Also, the new FFT benchmark ported from Accelerate gets even more memory savings than canny (>1.7x reduction in peak memory usage with futhark-opencl), so that's nice.

**Comment by nqpz** [comment]
We probably shouldn't merge it without first enabling memory block merging, though, since all the `structure cpu` and `structure gpu` annotations in the memory block merging tests will fail without the optimisation enabled.

**Comment by Athas** [comment]
Which of your branches should I try out if I want to take a look at some of the above?

**Comment by Athas** [comment]
I assume it's niels-memory-block-merging-final.

**Comment by nqpz** [comment]
Correct.

**Comment by Athas** [comment]
The problem with `tpacf` is not due to simplification.  Memory block merging produces this pattern:

    {mem(bytes_6976) step_two_kernel_result_mem_7005, i64 new_memory_size_7181; [binop_x_6956]i32 step_two_kernel_result_5973}

This is not a sensible pattern, as `new_memory_size_7181` is not actually used for the memory block.  This then confuses a simplification rule.  There is a secondary bug, which is that the internal type checker is not very diligent about checking the pattern for validity.  I can certainly improve that.

**Comment by Athas** [comment]
Actually, the problem is that the pattern does not match the type of the `if`-expression it is connected to.  It is OK for patterns to contain unused existential parts.

**Comment by Athas** [comment]
I have changed the type checking so that it is now clear that you are at fault.  You're welcome.

**Comment by Athas** [comment]
Also, you are really making your work look worse than it is at http://napoleon.hiperfit.dk/~niels/memory-block-merging/gpu01-2018-02-08-futhark-opencl/summary.htm

The problem is that you are letting slowdowns on tiny datasets have way too much impact.  When computing the impact on a benchmark, I recommend *summing the total combined runtime for all data sets* and using *that* to compute impact.  Right now I think you are taking the average of the impact per dataset, or something like that.

**Comment by nqpz** [comment]
I have fixed the pattern problem and verified that it doesn't cause errors with any other programs.

Also, here are some new measurements from a different machine: http://napoleon.hiperfit.dk/~niels/memory-block-merging/niels-laptop-2018-02-16-futhark-opencl/summary.html

The optimisation is still set to only run if the proper environment variables are set.  Should I enable it by default and give you a squashed commit?

**Comment by Athas** [comment]
You may squash it for merging, but don't enable it by default.  I would like to perform my own tests as well.

**Comment by nqpz** [comment]
Also note that I don't really trust my laptop for precise benchmarking.

**Comment by nqpz** [comment]
My main input is this: I think it would be a good priority to remove unnecessary Copy statements from the ImpGen representation, as described in the Runtime regressions section in this issue. I did this uglily in a a branch at some point, and it seems an easy win: If the destination and source are equal in both memory block name, offset, and (to some extent?) length, then a Copy should not be inserted. Or is this already handled somewhere?

Also, I wrote some tests at some point whose required features never got implemented; see http://napoleon.hiperfit.dk/~niels/memory-block-merging/tests-wip/. Most of them are extreme edge cases (I think), but some of them, like `coalescing/loop/ixfun-analysis-across-iterations.fut`, actually seem to represent real programs, and might make sense to get working at some point.
